Gems/Multiplayer/Code/Source/NetworkEntity/EntityReplication/EntityReplicator.cpp

```diff
@@ -465,38 +534,76 @@ namespace Multiplayer
 
     NetworkEntityUpdateMessage EntityReplicator::GenerateUpdatePacket()
     {
-        if (IsMarkedForRemoval() && OwnsReplicatorLifetime()) // TODO: clean this up
+        auto message = m_propertyPublisher->GenerateUpdatePacket(m_netBindComponent, WasMigrated());
+        if (message.GetIsDelete())
         {
-            // If the remote replicator is not established, we need to take ownership of the entity
-            AZLOG
-            (
+            AZLOG(
                 NET_RepDeletes,
-                "Sending delete replicator id %llu migrated %d to remote host %s",
+                "Sending delete replicator id %llu migrated=%s to remote host %s",
                 aznumeric_cast<AZ::u64>(GetEntityHandle().GetNetEntityId()),
-                WasMigrated() ? 1 : 0,
-                m_replicationManager.GetRemoteHostId().GetString().c_str()
-            );
-            return NetworkEntityUpdateMessage(GetEntityHandle().GetNetEntityId(), WasMigrated());
+                WasMigrated() ? "true" : "false",
+                m_replicationManager.GetRemoteHostId().GetString().c_str());
         }
+        return message;
+    }
 
-        NetBindComponent* netBindComponent = GetNetBindComponent();
-        const bool sendSliceName = !m_propertyPublisher->IsRemoteReplicatorEstablished();
+    EntityMigrationMessage EntityReplicator::GenerateMigrationPacket()
+    {
+        return m_propertyPublisher->GenerateMigrationPacket(m_netBindComponent);
+    }
 
-        NetworkEntityUpdateMessage updateMessage(GetRemoteNetworkRole(), GetEntityHandle().GetNetEntityId());
-        if (sendSliceName)
-        {
-            updateMessage.SetPrefabEntityId(netBindComponent->GetPrefabEntityId());
-        }
+    bool EntityReplicator::IsReadyToPublish() const
+    {
+        return (m_propertyPublisher != nullptr);
+    }
+
+    bool EntityReplicator::IsRemoteReplicatorEstablished() const
+    {
+        return m_propertyPublisher->IsRemoteReplicatorEstablished();
+    }
+
+    bool EntityReplicator::HasChangesToPublish()
+    {
+        return m_propertyPublisher->RequiresSerialization();
+    }
 
-        InputSerializer inputSerializer(updateMessage.ModifyData().GetBuffer(), static_cast<uint32_t>(updateMessage.ModifyData().GetCapacity()));
-        m_propertyPublisher->UpdateSerialization(inputSerializer);
-        updateMessage.ModifyData().Resize(inputSerializer.GetSize());
+    void EntityReplicator::SetRebasing()
+    {
+        AZ_Assert(m_propertyPublisher, "Expected to have a property publisher");
+        m_propertyPublisher->SetRebasing();
```

Comment:
Is it worth being a bit more defensive and null-guarding the actual implementation calls in case a replicator gets in a bad state? It might allow execution to continue for a time, albeit incorrectly and with unpredictable side-effects.

Reply:
yeah, wouldn't hurt. i'll make sure they all assert as well, so that way the problem is pretty obvious if it happens. it's definitely an API misuse if something tries to use them when the publisher or subscriber is null, but since it's state dependent and lifetime dependent, it's easy to see how someone could accidentally misuse the API. Guarded!